File: backend/app/api/recommend.py
# =============================================================================
# 智能零售用户行为分析系统 - 推荐系统模块 API 路由
# =============================================================================
# 4个接口:
#   8.1 GET  /recommend/hot           - 热门商品推荐（Redis缓存）
#   8.2 GET  /recommend/association   - 关联规则推荐（根据商品）
#   8.3 GET  /recommend/personalized  - 个性化推荐（协同过滤）
#   8.4 POST /recommend/refresh       - 刷新推荐缓存
# =============================================================================

from fastapi import APIRouter, Depends, Query
from sqlalchemy import func
from sqlalchemy.orm import Session
from typing import Dict, Optional
import json
from collections import defaultdict

from app.schemas.recommend import HotRecommendItem, AssociationRecommendItem, PersonalizedRecommendItem
from app.models.product import Product
from app.models.order import Order
from app.models.order_item import OrderItem
from app.models.behavior import Behavior
from app.models.base import get_db
from app.utils.response_utils import success_response
from app.utils.jwt_utils import get_current_user
from app.utils.redis_utils import cache_get, cache_set, cache_delete
from app.services.algorithm_outputs import (
    clear_algorithm_output_cache,
    get_algorithm_association_recommendations,
    get_algorithm_hot_recommendations,
    get_algorithm_personalized_recommendations,
)


router = APIRouter(prefix="/recommend", tags=["推荐系统"])


# ======================== 8.1 热门商品推荐 ========================

@router.get("/hot", summary="热门商品推荐")
async def get_hot_recommend(
    limit: int = Query(default=10, ge=1, le=50, description="返回数量"),
    db: Session = Depends(get_db),
    current_user: Dict = Depends(get_current_user)
) -> Dict:
    """
    获取热门商品推荐列表
    
    优先从Redis缓存读取，缓存键: recommend_hot_{limit}
    """
    cache_key = f"recommend_hot_{limit}"
    cached = await cache_get(cache_key)
    if cached:
        return success_response(data=json.loads(cached))

    algorithm_data = get_algorithm_hot_recommendations(db, limit)
    if algorithm_data:
        await cache_set(cache_key, json.dumps(algorithm_data, ensure_ascii=False), 300)
        return success_response(data=algorithm_data)
    
    products = db.query(Product).filter(
        Product.deleted_at == None, Product.status == "on"
    ).order_by(Product.sales_count.desc()).limit(limit).all()
    
    data = [
        HotRecommendItem(
            product_id=p.id, name=str(p.name), price=float(p.price),
            sales_count=p.sales_count or 0, rating=float(p.rating or 0)
        ).model_dump() for p in products
    ]
    await cache_set(cache_key, json.dumps(data, ensure_ascii=False), 300)
    return success_response(data=data)


# ======================== 8.2 关联规则推荐 ========================

@router.get("/association", summary="关联规则推荐")
async def get_association_recommend(
    product_id: int = Query(..., description="当前商品ID"),
    limit: int = Query(default=10, ge=1, le=20, description="返回数量"),
    db: Session = Depends(get_db),
    current_user: Dict = Depends(get_current_user)
) -> Dict:
    """
    根据指定商品获取关联推荐
    
    基于订单共购关系查询关联商品。

    真实数据量较大时，实时 Apriori 会阻塞后端进程；这里改为轻量 SQL 聚合：
    1. 找出包含当前商品的订单
    2. 统计这些订单里共同出现的其他商品
    3. 无结果时用热门商品兜底
    """
    cache_key = f"recommend_association_v2_{product_id}_{limit}"
    cached = await cache_get(cache_key)
    if cached:
        return success_response(data=json.loads(cached))

    algorithm_data = get_algorithm_association_recommendations(db, product_id, limit)
    if algorithm_data:
        await cache_set(cache_key, json.dumps(algorithm_data, ensure_ascii=False), 300)
        return success_response(data=algorithm_data)

    target_order_ids = (
        db.query(OrderItem.order_id)
        .filter(OrderItem.product_id == product_id)
        .subquery()
    )
    target_order_count = (
        db.query(func.count(func.distinct(OrderItem.order_id)))
        .filter(OrderItem.product_id == product_id)
        .scalar()
        or 0
    )
    total_order_count = db.query(func.count(Order.id)).scalar() or 1

    rows = (
        db.query(
            Product.id,
            Product.name,
            Product.price,
            Product.sales_count,
            func.count(func.distinct(OrderItem.order_id)).label("co_count"),
        )
        .join(OrderItem, OrderItem.product_id == Product.id)
        .filter(OrderItem.order_id.in_(target_order_ids))
        .filter(Product.id != product_id)
        .filter(Product.deleted_at == None)
        .filter(Product.status == "on")
        .group_by(Product.id, Product.name, Product.price)
        .order_by(func.count(OrderItem.id).desc(), Product.sales_count.desc())
        .limit(limit)
        .all()
    )

    row_product_ids = [row.id for row in rows]
    product_order_counts = {}
    if row_product_ids:
        product_order_counts = dict(
            db.query(
                OrderItem.product_id,
                func.count(func.distinct(OrderItem.order_id)),
            )
            .filter(OrderItem.product_id.in_(row_product_ids))
            .group_by(OrderItem.product_id)
            .all()
        )

    data = []
    for row in rows:
        confidence = float(row.co_count) / float(target_order_count) if target_order_count else 0.0
        other_order_count = product_order_counts.get(row.id, row.co_count)
        other_probability = float(other_order_count) / float(total_order_count) if total_order_count else 0.0
        lift = confidence / other_probability if other_probability else 1.0
        data.append(
            AssociationRecommendItem(
                product_id=row.id,
                name=str(row.name),
                price=float(row.price),
                confidence=round(min(confidence, 1.0), 4),
                lift=round(max(lift, 1.0), 2),
            ).model_dump()
        )

    if not data:
        hot_products = (
            db.query(Product)
            .filter(Product.deleted_at == None, Product.status == "on", Product.id != product_id)
            .order_by(Product.sales_count.desc())
            .limit(limit)
            .all()
        )
        max_sales = max([p.sales_count or 0 for p in hot_products] or [1]) or 1
        data = []
        for p in hot_products:
            confidence = float(p.sales_count or 0) / float(max_sales)
            data.append(
                AssociationRecommendItem(
                    product_id=p.id,
                    name=str(p.name),
                    price=float(p.price),
                    confidence=round(confidence, 4),
                    lift=round(1.0 + confidence, 2),
                ).model_dump()
            )

    await cache_set(cache_key, json.dumps(data, ensure_ascii=False), 300)
    return success_response(data=data)


# ======================== 8.3 个性化推荐 ========================

@router.get("/personalized", summary="个性化推荐")
async def get_personalized_recommend(
    user_id: int = Query(..., description="用户ID"),
    limit: int = Query(default=10, ge=1, le=20, description="返回数量"),
    db: Session = Depends(get_db),
    current_user: Dict = Depends(get_current_user)
) -> Dict:
    """
    获取个性化推荐（轻量规则版）
    
    推荐策略：
    1. 找出用户购买过的商品分类
    2. 在同分类中推荐高销量商品（排除已购买）
    3. 如果不够，补充全局热门商品
    4. 附加推荐理由
    """
    cache_key = f"recommend_personalized_v2_{user_id}_{limit}"
    cached = await cache_get(cache_key)
    if cached:
        return success_response(data=json.loads(cached))

    algorithm_data = get_algorithm_personalized_recommendations(db, user_id, limit)
    if algorithm_data:
        await cache_set(cache_key, json.dumps(algorithm_data, ensure_ascii=False), 300)
        return success_response(data=algorithm_data)

    # 1. 获取用户已购商品
    bought = set()
    orders = db.query(Order).filter(
        Order.user_id == user_id, Order.status.in_(["completed", "paid", "shipped"])
    ).all()
    for o in orders:
        for item in o.items:
            bought.add(item.product_id)
    
    # 2. 获取用户偏好分类
    cat_ids = set()
    for pid in bought:
        p = db.query(Product).filter(Product.id == pid).first()
        if p and p.category_id:
            cat_ids.add(p.category_id)
    
    # 3. 在偏好分类中推荐
    result = []
    seen = set(bought)
    
    for cid in cat_ids:
        candidates = db.query(Product).filter(
            Product.category_id == cid, Product.deleted_at == None,
            Product.status == "on", ~Product.id.in_(bought)
        ).order_by(Product.sales_count.desc()).limit(3).all()
        for p in candidates:
            if p.id not in seen:
                seen.add(p.id)
                result.append(PersonalizedRecommendItem(
                    product_id=p.id, name=str(p.name), price=float(p.price),
                    score=round(float(p.sales_count or 0) / 10000, 2),
                    reason="根据您的购买偏好推荐"
                ).model_dump())
    
    # 4. 补充热门商品
    if len(result) < limit:
        hot = db.query(Product).filter(
            Product.deleted_at == None, Product.status == "on",
            ~Product.id.in_(seen)
        ).order_by(Product.sales_count.desc()).limit(limit - len(result)).all()
        for p in hot:
            seen.add(p.id)
            result.append(PersonalizedRecommendItem(
                product_id=p.id, name=str(p.name), price=float(p.price),
                score=round(float(p.sales_count or 0) / 10000, 2),
                reason="热门商品推荐"
            ).model_dump())
    
    data = result[:limit]
    await cache_set(cache_key, json.dumps(data, ensure_ascii=False), 300)
    return success_response(data=data)


# ======================== 8.4 刷新推荐结果 ========================

@router.post("/refresh", summary="刷新推荐结果")
async def refresh_recommend(
    current_user: Dict = Depends(get_current_user)
) -> Dict:
    """
    清除推荐相关Redis缓存，下次访问时自动重新计算
    
    清除的缓存键前缀: recommend_*, hot_products_*
    """
    # 清除热门推荐缓存
    for i in [5, 10, 20, 50]:
        await cache_delete(f"recommend_hot_{i}")
        await cache_delete(f"hot_products_sales_{i}")
        await cache_delete(f"hot_products_rating_{i}")
    
    # 清除常见推荐缓存。Redis 工具当前只支持精确 key，这里清理常用数量的缓存。
    # 个性化/关联推荐的其他 key 会在 300 秒后自动过期。
    for i in [5, 10, 20]:
        await cache_delete(f"recommend_association_1_{i}")

    clear_algorithm_output_cache()
    
    return success_response(data={"refreshed": True}, message="推荐缓存已刷新")

File: frontend/src/api/recommend.ts
// =====================================================
// src/api/recommend.ts
// 推荐系统模块 API 接口层
// 严格遵循 API 接口规范文档 v1.0 第八章
// 包含: 热门推荐、关联推荐、个性化推荐、刷新推荐
// 当前使用 Mock 数据，后端就绪后设置 USE_MOCK = false
// =====================================================

import request from '@/utils/request'
import type { ApiResponse, RecommendItem } from '@/types/api'

const USE_MOCK = false

// ==================== Mock 数据 ====================

const hotProducts: RecommendItem[] = [
  { product_id: 1, name: '无线蓝牙耳机 Pro', price: 299.00, sales_count: 1256, rating: 4.8 },
  { product_id: 6, name: '机械键盘 RGB', price: 299.00, sales_count: 1089, rating: 4.7 },
  { product_id: 2, name: '智能手表 S3', price: 899.00, sales_count: 987, rating: 4.6 },
  { product_id: 4, name: '便携充电宝 20000mAh', price: 129.00, sales_count: 856, rating: 4.5 },
  { product_id: 13, name: '运动跑鞋', price: 499.00, sales_count: 743, rating: 4.4 },
  { product_id: 3, name: 'Type-C 数据线 1m', price: 29.90, sales_count: 689, rating: 4.3 },
  { product_id: 5, name: '降噪耳机罩', price: 399.00, sales_count: 612, rating: 4.5 },
  { product_id: 10, name: '高清摄像头 1080P', price: 249.00, sales_count: 534, rating: 4.2 },
  { product_id: 16, name: '智能台灯', price: 159.00, sales_count: 478, rating: 4.6 },
  { product_id: 7, name: '无线鼠标', price: 129.00, sales_count: 423, rating: 4.1 },
]

const associationRecs: Record<number, RecommendItem[]> = {
  1: [
    { product_id: 4, name: '便携充电宝 20000mAh', price: 129.00, confidence: 0.65, lift: 2.3 },
    { product_id: 3, name: 'Type-C 数据线 1m', price: 29.90, confidence: 0.58, lift: 1.9 },
    { product_id: 8, name: 'USB 集线器 7口', price: 69.00, confidence: 0.42, lift: 1.6 },
  ],
  2: [
    { product_id: 6, name: '机械键盘 RGB', price: 299.00, confidence: 0.52, lift: 2.1 },
    { product_id: 7, name: '无线鼠标', price: 129.00, confidence: 0.48, lift: 1.8 },
    { product_id: 9, name: '平板电脑支架', price: 79.00, confidence: 0.38, lift: 1.5 },
  ],
  6: [
    { product_id: 7, name: '无线鼠标', price: 129.00, confidence: 0.78, lift: 3.1 },
    { product_id: 9, name: '平板电脑支架', price: 79.00, confidence: 0.61, lift: 2.4 },
    { product_id: 2, name: '智能手表 S3', price: 899.00, confidence: 0.35, lift: 1.4 },
  ],
}

const personalizedRecs: Record<number, RecommendItem[]> = {
  1: [
    { product_id: 6, name: '机械键盘 RGB', price: 299.00, score: 0.92, reason: '您浏览过类似商品' },
    { product_id: 10, name: '高清摄像头 1080P', price: 249.00, score: 0.87, reason: '与您购买过的商品相关' },
    { product_id: 2, name: '智能手表 S3', price: 899.00, score: 0.81, reason: '根据您的消费偏好推荐' },
    { product_id: 16, name: '智能台灯', price: 159.00, score: 0.75, reason: '同类用户也购买了' },
  ],
  5: [
    { product_id: 13, name: '运动跑鞋', price: 499.00, score: 0.89, reason: '与您浏览过的商品相关' },
    { product_id: 25, name: '瑜伽垫', price: 79.00, score: 0.83, reason: '根据您的消费偏好推荐' },
    { product_id: 27, name: '登山背包', price: 399.00, score: 0.76, reason: '同类用户也购买了' },
  ],
}

const defaultPersonalized: RecommendItem[] = [
  { product_id: 1, name: '无线蓝牙耳机 Pro', price: 299.00, score: 0.85, reason: '热门商品推荐' },
  { product_id: 4, name: '便携充电宝 20000mAh', price: 129.00, score: 0.78, reason: '高评分商品推荐' },
  { product_id: 13, name: '运动跑鞋', price: 499.00, score: 0.72, reason: '新上架商品推荐' },
]

const mockDelay = <T>(data: T, d = 300): Promise<T> =>
  new Promise((r) => setTimeout(() => r(data), d))

// ==================== API 接口函数 ====================

/**
 * 8.1 热门商品推荐
 * 接口: GET /recommend/hot
 * 描述: 获取热门商品推荐列表（Redis 缓存）
 */
export const getHotRecommend = (limit = 10): Promise<ApiResponse<RecommendItem[]>> => {
  if (USE_MOCK) return mockDelay({ code: 200, message: 'success', data: hotProducts.slice(0, limit) })
  return request.get('/recommend/hot', { params: { limit } })
}

/**
 * 8.2 关联规则推荐
 * 接口: GET /recommend/association
 * 描述: 根据指定商品获取关联推荐（基于 Apriori 算法）
 */
export const getAssociationRecommend = (
  productId: number,
  limit = 10
): Promise<ApiResponse<RecommendItem[]>> => {
  if (USE_MOCK) {
    const recs = associationRecs[productId] || Object.values(associationRecs).flat().slice(0, 3).map((r, i) => ({ ...r, product_id: r.product_id + i }))
    return mockDelay({ code: 200, message: 'success', data: recs.slice(0, limit) })
  }
  return request.get('/recommend/association', { params: { product_id: productId, limit } })
}

/**
 * 8.3 个性化推荐
 * 接口: GET /recommend/personalized
 * 描述: 获取指定用户的个性化推荐（协同过滤/基于内容）
 */
export const getPersonalizedRecommend = (
  userId: number,
  limit = 10
): Promise<ApiResponse<RecommendItem[]>> => {
  if (USE_MOCK) {
    const recs = personalizedRecs[userId] || defaultPersonalized
    return mockDelay({ code: 200, message: 'success', data: recs.slice(0, limit) })
  }
  return request.get('/recommend/personalized', { params: { user_id: userId, limit } })
}

/**
 * 8.4 刷新推荐结果
 * 接口: POST /recommend/refresh
 * 描述: 触发后台重新计算所有推荐结果
 */
export const refreshRecommend = (): Promise<ApiResponse<null>> => {
  if (USE_MOCK) {
    return mockDelay({ code: 200, message: '推荐结果已重新计算', data: null }, 1500)
  }
  return request.post('/recommend/refresh')
}
